src/Compilers/VisualBasic/Portable/Symbols/Tuples/TupleTypeSymbol.vb

```diff
@@ -0,0 +1,987 @@
+                elementTypes = instance.ToImmutableAndFree()
+            Else
+                elementTypes = tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics
+            End If
+
+            Return New TupleTypeSymbol(locations, tupleCompatibleType, elementLocations, elementNames, elementTypes)
+        End Function
+
+        Private Shared Function EnsureRestExtensionsAreTuples(tupleCompatibleType As NamedTypeSymbol) As NamedTypeSymbol
+            If Not tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics(TupleTypeSymbol.RestPosition - 1).IsTupleType Then
+                Dim nonTupleTypeChain As ArrayBuilder(Of NamedTypeSymbol) = ArrayBuilder(Of NamedTypeSymbol).GetInstance()
+                Dim namedTypeSymbol As NamedTypeSymbol = tupleCompatibleType
+
+                Do
+                    nonTupleTypeChain.Add(namedTypeSymbol)
+                    namedTypeSymbol = CType(namedTypeSymbol.TypeArgumentsNoUseSiteDiagnostics(TupleTypeSymbol.RestPosition - 1), NamedTypeSymbol)
+                Loop While namedTypeSymbol.Arity = TupleTypeSymbol.RestPosition
+
+                If Not namedTypeSymbol.IsTupleType Then
+                    nonTupleTypeChain.Add(namedTypeSymbol)
+                End If
+
+                Debug.Assert(nonTupleTypeChain.Count > 1)
+                tupleCompatibleType = nonTupleTypeChain.Pop()
+
+                Dim typeArgumentsBuilder As ArrayBuilder(Of TypeWithModifiers) = ArrayBuilder(Of TypeWithModifiers).GetInstance(TupleTypeSymbol.RestPosition)
+                Do
+                    Dim extensionTuple As TupleTypeSymbol = TupleTypeSymbol.Create(CType(Nothing, Location), tupleCompatibleType, Nothing, Nothing)
+                    tupleCompatibleType = nonTupleTypeChain.Pop()
+                    tupleCompatibleType = TupleTypeSymbol.ReplaceRestExtensionType(tupleCompatibleType, typeArgumentsBuilder, extensionTuple)
+                Loop While nonTupleTypeChain.Count <> 0
+
+                typeArgumentsBuilder.Free()
+                nonTupleTypeChain.Free()
+            End If
+            Return tupleCompatibleType
+        End Function
+
+        Private Shared Function ReplaceRestExtensionType(tupleCompatibleType As NamedTypeSymbol, typeArgumentsBuilder As ArrayBuilder(Of TypeWithModifiers), extensionTuple As TupleTypeSymbol) As NamedTypeSymbol
+            Dim modifiers As ImmutableArray(Of ImmutableArray(Of CustomModifier)) = Nothing
+            Dim hasTypeArgumentsCustomModifiers As Boolean = tupleCompatibleType.HasTypeArgumentsCustomModifiers
+
+            If hasTypeArgumentsCustomModifiers Then
+                modifiers = tupleCompatibleType.TypeArgumentsCustomModifiers
+            End If
+
+            Dim typeArgumentsNoUseSiteDiagnostics As ImmutableArray(Of TypeSymbol) = tupleCompatibleType.TypeArgumentsNoUseSiteDiagnostics
+            typeArgumentsBuilder.Clear()
+
+            For i As Integer = 0 To TupleTypeSymbol.RestPosition - 1 - 1
```

Comment:
I think   -1 -1  is better here.   The last "-1" is just because this is a VB For, so it is more obvious that we iterate  elements -1

---
In reply to: [71555341](https://github.com/dotnet/roslyn/pull/12596#discussion_r71555341) [](ancestors = 71555341)